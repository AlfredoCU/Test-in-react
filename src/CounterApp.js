import React, { useState } from "react";
import PropTypes from "prop-types";

const CounterApp = ({ value }) => {
  const [counter, setCounter] = useState(value);

  const handleAdd = () => {
    setCounter(counter + 1);
    // setCounter((c) => c + 1);
  };

  const handleSubtract = () => setCounter(counter - 1);
  const handleReset = () => setCounter(value);

  return (
    <>
      <h1>CounterApp</h1>
      <p>{counter}</p>
      <button onClick={handleAdd}>+1</button>
      <button onClick={handleReset}>Reset</button>
      <button onClick={handleSubtract}>-1</button>
    </>
  );
};

CounterApp.propTypes = {
  value: PropTypes.number,
};

CounterApp.defaultProps = {
  value: 0,
};

export default CounterApp;

/* 
  const handleAdd = (e) => {
    console.log(e);
    // No retorna nada.
  };
  
  Ejemplo de funcion:
  <button onClick={(e) => console.log(e)}>Incrementar</button>
  
  Sólo si es necesario pasar agumentos:
  <button onClick={(e) => handleAdd(e)}>Incrementar</button>
  Pero podemos reducir el código sólo si cumple con lo siguiente o sino enviaremos argumentos ->

  Tenemos una función de flecha, el primer argumento que recibe es enviado como primer argumento 
  a una función que está adentro.
  Por consecuencia, podríamos borrar esto y dejar únicamente la referencia a la función que yo quiero ejecutar.
  
  Pasamos la referencia de la función para que esta sea ejectuda:
  <button onClick={handleAdd}>Incrementar</button>

  Otro caso retorna un undefined y no funciona, para que funcione la funcion handleAdd debe de retornar una función:
  <button onClick={handleAdd()}>Incrementar</button>
  Aparte esto hace que se ejecute la función cuando se renderiza, ignorando el onClick, esto sólo debe de ser usado
  cuando es necesario ejecutar una función de forma automatica.
*/
