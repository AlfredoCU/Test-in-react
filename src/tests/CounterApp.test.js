import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import CounterApp from "../CounterApp";

describe("Prueba a <CounterApp />", () => {
  //* INICIALIZA WRAPPER ANTES DE CADA PRUEBA.
  let wrapper = shallow(<CounterApp />);
  beforeEach(() => {
    wrapper = shallow(<CounterApp />);
  });

  test("Prueba a mostrar correctamente la estructura", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar el valor a 100", () => {
    const v = 100;
    const wrapper = shallow(<CounterApp value={v} />);
    // Puedes poner id o class tambiÃ©n.
    const valor = wrapper.find("p").text(); // trim(); borra los cambios
    expect(valor).toBe(`${v}`);
  });

  //! LAS PRUEBAS VAN EN SECUENCIA.

  test("Prueba al btn +1", () => {
    wrapper.find("button").at(0).simulate("click");
    const valor = wrapper.find("p").text();
    expect(valor).toBe("1");
    // console.log(btn.html()); Muestra el html del btn.
  });

  test("Prueba al btn reset", () => {
    wrapper.find("button").at(1).simulate("click");
    const valor = wrapper.find("p").text();
    expect(valor).toBe("0");
  });

  test("Prueba al btn -1", () => {
    wrapper.find("button").at(2).simulate("click");
    const valor = wrapper.find("p").text();
    expect(valor).toBe("-1");
  });
});

//* SE PUEDE REALIZAR LAS PRUEBAS DE ESTA FORMA TAMBIEN.
// test('Debe de colocar el valor por defecto con el btn reset', () => {
//   const wrapper = shallow( <CounterApp value={ 105 } /> );
//   wrapper.find('button').at(0).simulate('click');
//   wrapper.find('button').at(0).simulate('click');
//   wrapper.find('button').at(1).simulate('click');
//   const counterText = wrapper.find('h2').text().trim();
//   expect( counterText ).toBe('105');
// })
