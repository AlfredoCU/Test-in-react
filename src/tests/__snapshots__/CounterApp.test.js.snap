// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba a <CounterApp /> Prueba a mostrar correctamente la estructura 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <p>
    0
  </p>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
</Fragment>
`;
